Performance Scenario Cleanup

jjohannes opened the pull request.

oehme commented:
Deriving the test project from the first word of the method name is too much magic in my opinion. I think we should remove that again.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaFirstUsePerformanceTest.groovy:
```diff
@@ -19,16 +19,14 @@ package org.gradle.performance.regression.java
-        runner.args = ['--recompile-scripts']
+        runner.tasksToRun = ['tasks']
+        runner.gradleOpts = ["-Xms${memory}", "-Xmx${memory}"]
+        runner.args = ['--recompile-scripts'] // This is an approximation of first use: we recompile the scripts
```
Should be running without the daemon

jjohannes replied:
Done.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/corefeature/TasksReportPerformanceTest.groovy:
```diff
@@ -1,72 +0,0 @@
-import static org.gradle.performance.regression.corefeature.TasksReportPerformanceTest.TasksReport.BASIC_TASKS_REPORT
-import static org.gradle.performance.regression.corefeature.TasksReportPerformanceTest.TasksReport.DETAILED_TASKS_REPORT
-
-class TasksReportPerformanceTest extends AbstractCrossVersionPerformanceTest {
```
We should still be testing this, it is different from "first use" and we used to have lots of problems with the `tasks` task. We don't want to regress there again.

jjohannes replied:
Re-added `JavaTasksPerformanceTest`

oehme commented on subprojects/performance/templates.gradle:
```diff
@@ -20,90 +20,65 @@ import org.gradle.testing.performance.generator.tasks.NativeProjectGeneratorTask
-
-task largeWithJUnit(type: JvmProjectGeneratorTask) {
+task largeMultiProjectJava(type: JvmProjectGeneratorTask) {
     projects = 50
```
That's one order of magnitude too small.

jjohannes replied:
Done.

oehme commented on subprojects/performance/templates.gradle:
```diff
@@ -22,7 +22,7 @@ import org.gradle.testing.performance.generator.tasks.RemoteProject
 // === Java ===
 task largeMonolithicProjectJava(type: JvmProjectGeneratorTask, description: 'Generates a single project with lots of source files') {
-    sourceFiles = 50000
+    sourceFiles = 15000
```
50000 was good, that's a realistic size for such a project I think

jjohannes replied:
Changed back.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaFirstUsePerformanceTest.groovy:
```diff
@@ -26,7 +26,7 @@ class JavaFirstUsePerformanceTest extends AbstractCrossVersionPerformanceTest {
         runner.tasksToRun = ['tasks']
         runner.gradleOpts = ["-Xms${memory}", "-Xmx${memory}"]
-        runner.args = ['--recompile-scripts'] // This is an approximation of first use: we recompile the scripts
+        runner.args = ['--recompile-scripts', '--no-daemon'] // This is an approximation of first use: we recompile the scripts
```
Should use runner.useDaemon=false, this is inspected by other parts of the infrastructure

jjohannes replied:
Changed. Wasn't aware of that.

jjohannes commented:
> Deriving the test project from the first word of the method name is too much magic in my opinion. I think we should remove that again.

This was really more about the `testId` - I wanted the test project to be part of it. The tests are made up of two dimensions. The scenario (e.g. `CleanAssemble`) and the test project (e.g. `largeMonolithicProjectJava`). It would be good if all tests could be named/categorised following this. So in the report, or if you want to run an individual test scenario, it would be great if you only need to know these two. So you clearly see what a test is, instead of having inconsistent `testID` patterns as we have now. That's why I think it is good to have both dimensions in the `testID` in the end. 

This is a way to enforce this pattern. (Or at least strongly suggest it, as you can still define `testId` and `testProject` manually.)

I also don't like repeating the testProject twice like it was done in many tests before:
```
def "something #testProject something else (something we better mention here)" {
   runner.testProject = testProject // <- this feels like unnecessary boilerplate
}
```

Maybe there is a less-magical way to do this I have not thought of.

> Do these example projects have plenty of external/project dependencies? If not, we need to add those.

Depends on the definition of _plenty_ :) Currently each (sub-)project has these dependencies:
https://github.com/gradle/gradle/blob/996dda7441496b0835962fff622da9ef7d7495fb/subprojects/build-scan-performance/src/templates/project-with-source/build.gradle#L31-L37

> We still need to test no-daemon performance.

I have added `--no-daemon` to the _first use_ scenario. Do we need more with `--no-daemon`? I think we do not want to run all scenarios with `--no-daemon` to not blow up the number of tests? Or should we do something like adding them to the experimental runs (like the parallel test)?

> Afaik there is no more test that checks our (mostly) up-to-date behavior. We should have the up-to-date scenario at least until we have the incremental change scenarios

Right. I added an UpToDate scenario that we might replace with the incremental scenario once we have that.

oehme commented:
> I have added --no-daemon to the first use scenario. 

That should be enough.

> Depends on the definition of plenty :) Currently each (sub-)project has these dependencies:
https://github.com/gradle/gradle/blob/996dda7441496b0835962fff622da9ef7d7495fb/subprojects/build-scan-performance/src/templates/project-with-source/build.gradle#L31-L37

Those are only external dependencies, but we also need project dependencies. (see if-statement below)

Regarding the naming and test project: I think a small bit of boilerplate is fine. We want the test names to be readable (they are currently all over the place), but I think we can create consistency through setting good examples. If we see a pattern emerging at some point, we can automate that, but I wouldn't start with that.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/experiment/java/ParallelBuildPerformanceTest.groovy:
```diff
@@ -44,8 +44,8 @@ class ParallelBuildPerformanceTest extends AbstractCrossBuildPerformanceTest {
         testProject                  | memory | warmUpRuns | runs
-        "largeMonolithicProjectJava" | '608m' | 2          | 6
-        "largeMultiProjectJava"      | '608m' | 2          | 6
+        "largeMonolithicJavaProject" | '4g'   | 2          | 6
```
For such large projects we should use forked compilation instead and keep the main daemon small.

jjohannes replied:
Done.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaIDEModelPerformanceTest.groovy:
```diff
@@ -135,11 +134,9 @@ class ToolingApiIdeModelCrossVersionPerformanceTest extends AbstractToolingApiCr
-        "bigOldJava"        | '608m'
-        "lotDependencies"   | '256m'
+        testProject                  | memory| warmUpRuns | runs
+        "largeMonolithicJavaProject" | '2g'  | 4          | 10
```
Should get rid of memory arguments like in the other tests

jjohannes replied:
Done.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaNonABIChangePerformanceTest.groovy:
```diff
@@ -0,0 +1,44 @@
+        runner.testProject = testProject
+        runner.tasksToRun = ['assemble']
+        runner.targetVersions = ["3.5-20170221000043+0000"]
+        runner.addBuildExperimentListener(new JavaSourceFileUpdater(1))
```
Should replace this "source file updater" with the two updaters we have in gradle-profiler (abiChange/nonAbiChange, taking the relative path of the file to change)

jjohannes replied:
Done - copied and from `gradle-profiler` + slight adjustments. Might further refine them together with replacing the generator.

oehme commented on subprojects/performance/src/templates/root-project/gradle.properties:
```diff
@@ -0,0 +1,2 @@
+<% if (binding.hasVariable("daemonMemory")) { %>org.gradle.jvmargs=-Xmxs${binding.getVariable("daemonMemory")} -Xmx${binding.getVariable("daemonMemory")}<% } %>
```
We should probably add `compilerMemory` and `testRunnerMemory` here instead of inline in the build script, so they are easier to see/change at a glance.

jjohannes replied:
Done.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaIDEModelPerformanceTest.groovy:
```diff
@@ -32,7 +32,7 @@ class JavaIDEModelPerformanceTest extends AbstractToolingApiCrossVersionPerforma
             action {
                 def model = model(tapiClass(EclipseProject))
-                    .setJvmArguments(customizeJvmOptions(["-Xms$memory", "-Xmx$memory"])).get()
+                    .setJvmArguments(customizeJvmOptions()).get()
```
I don't think we actually need to set them here. This call looks confusing :)

jjohannes replied:
I was afraid to break some unexpected implicit behaviour. :) - Remove.

oehme commented on subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/PerformanceTestJvmOptions.groovy:
```diff
@@ -22,7 +22,7 @@ import org.gradle.api.JavaVersion
 
 @CompileStatic
 class PerformanceTestJvmOptions {
```
I think this class is now obsolete.

jjohannes replied:
It is still used by some non-Java tests (e.g. `NativeBuildPerformanceTest`). I would leave it in until we refactored those as well.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaTestChangePerformanceTest.groovy:
```diff
@@ -39,8 +39,11 @@ class JavaTestChangePerformanceTest extends AbstractCrossVersionPerformanceTest
+        "largeJavaMultiProject"            | 2          | 6    | "project450/src/main/java/org/gradle/test/performance450_1/Production450_1.java"
+        "largeJavaMultiProjectTestNG"      | 2          | 6    | "project450/src/main/java/org/gradle/test/performance450_1/Production450_1.java"
+
+        //"largeMonolithicJavaProject" - there is no increment in the test running part here, so it would correspond to a source change + clean test
```
I'd probably change this comment to "We don't support incremental testing within a single test task, so this would be the same as the `cleanTest test` case.

jjohannes replied:
Sounds better. Replaced the comment.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaCleanTestPerformanceTest.groovy:
```diff
@@ -0,0 +1,46 @@
+        testProject                          | warmUpRuns | runs
+        "largeMonolithicJavaProject"         | 2          | 6
+        //"largeMonolithicJavaProjectTestNG" | 2          | 6 // testNG requires more test workers, which take too long to start up
+        //"largeJavaMultiProject"            | 2          | 6 // due to long test worker startup times, multi-project test runs take too long
```
We should probably have a medium-sized project for testing until we improve the test performance. Not testing at all feels wrong.

oehme commented on subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaTestChangePerformanceTest.groovy:
```diff
@@ -0,0 +1,49 @@
+        where:
+        testProject                        | warmUpRuns | runs | fileToChange
+        "largeJavaMultiProject"            | 2          | 6    | "project450/src/main/java/org/gradle/test/performance450_1/Production450_1.java"
+        "largeJavaMultiProjectTestNG"      | 2          | 6    | "project50/src/main/java/org/gradle/test/performance50_1/Production50_1.java"
```
This should probably be called "mediumJavaMultiProjectWithTestNG", because it has much less projects.

jjohannes replied:
Changed. Also added a java projects enum.
